Applet/src/com/android/javacard/keymaster/KMKeymasterApplet.java

```diff
@@ -3726,68 +3755,116 @@ private static void createEncryptedKeyBlob(byte[] scratchPad) {
     encryptSecret(scratchPad);
     // create key blob array
     KMArray.cast(data[KEY_BLOB]).add(KEY_BLOB_SECRET, data[SECRET]);
     KMArray.cast(data[KEY_BLOB]).add(KEY_BLOB_AUTH_TAG, data[AUTH_TAG]);
     KMArray.cast(data[KEY_BLOB]).add(KEY_BLOB_NONCE, data[NONCE]);
+    KMArray.cast(data[KEY_BLOB]).add(KEY_BLOB_VERSION, data[KEY_BLOB_VERSION_DATA_OFFSET]);
 
-    //TODO remove the following temporary creation of keyblob.
     short tempChar = KMKeyCharacteristics.instance();
     short emptyParam = KMArray.instance((short) 0);
     emptyParam = KMKeyParameters.instance(emptyParam);
     KMKeyCharacteristics.cast(tempChar).setStrongboxEnforced(data[SB_PARAMETERS]);
     KMKeyCharacteristics.cast(tempChar).setKeystoreEnforced(emptyParam);
     KMKeyCharacteristics.cast(tempChar).setTeeEnforced(data[TEE_PARAMETERS]);
     KMArray.cast(data[KEY_BLOB]).add(KEY_BLOB_PARAMS, tempChar);
-    //KMArray.cast(data[KEY_BLOB]).add(KEY_BLOB_PARAMS, data[KEY_CHARACTERISTICS]);
 
     // allocate reclaimable memory.
     short buffer = repository.alloc((short) 1024);
     short keyBlob = encoder.encode(data[KEY_BLOB], repository.getHeap(), buffer);
     data[KEY_BLOB] = KMByteBlob.instance(repository.getHeap(), buffer, keyBlob);
   }
 
-  private short parseEncryptedKeyBlob(short keyBlob, short appId, short appData, byte[] scratchPad) {
-    short parsedBlob = KMType.INVALID_VALUE;
-    short rot = readROT(scratchPad);
-    if (rot == KMType.INVALID_VALUE) {
-      KMException.throwIt(KMError.UNKNOWN_ERROR);
-    }
+  private short readKeyBlobVersion(short keyBlob) {
+    short version = KMType.INVALID_VALUE;
     try {
-      parsedBlob = decoder.decodeArray(keyBlob(),
-              KMByteBlob.cast(keyBlob).getBuffer(),
-              KMByteBlob.cast(keyBlob).getStartOff(),
-              KMByteBlob.cast(keyBlob).length());
-      if (KMArray.cast(parsedBlob).length() < 4) {
-        KMException.throwIt(KMError.INVALID_KEY_BLOB);
+      version = decoder.readKeyblobVersion(
+          KMByteBlob.cast(keyBlob).getBuffer(),
+          KMByteBlob.cast(keyBlob).getStartOff(),
+          KMByteBlob.cast(keyBlob).length());
+      if (version == KMType.INVALID_VALUE) {
+        // If Version is not present. Then it is either an old KeyBlob or
+        // corrupted KeyBlob.
+        version = 0;
       }
+      version = KMInteger.cast(version).getShort();
+    } catch(Exception e) {
+      KMException.throwIt(KMError.INVALID_KEY_BLOB);
+    }
+    return version;
+  }
 
-      // initialize data
-      data[SECRET] = KMArray.cast(parsedBlob).get(KEY_BLOB_SECRET);
-      data[NONCE]= KMArray.cast(parsedBlob).get(KEY_BLOB_NONCE);
-      data[AUTH_TAG] = KMArray.cast(parsedBlob).get(KEY_BLOB_AUTH_TAG);
-      data[KEY_CHARACTERISTICS] = KMArray.cast(parsedBlob).get(KEY_BLOB_PARAMS);
+  private void readKeyBlobParams(short version, short parsedKeyBlob) {
+    data[KEY_BLOB] = parsedKeyBlob;
+    // initialize data
+    if (version == 0) {
```

Comment:
use a switch case for version to make it future proof

Reply:
Added switch statement